plutus-chain-index-core/src/Plutus/ChainIndex.hs

```diff
@@ -54,24 +58,53 @@ runChainIndexEffects runReq action =
         $ handleChainIndexEffects runReq
         $ raiseEnd action
 
--- | Handle the chain index effects from the set of all effects.
+-- | Handle the chain index effects from the set of all effects using the
+-- SQLite store.
 handleChainIndexEffects
-    :: (LastMember IO effs, Member (LogMsg ChainIndexLog) effs)
+    :: forall effs a. (LastMember IO effs, Member (LogMsg ChainIndexLog) effs)
     => RunRequirements
     -> Eff (ChainIndexQueryEffect ': ChainIndexControlEffect ': BeamEffect ': effs) a
     -> Eff effs (Either ChainIndexError a)
-handleChainIndexEffects RunRequirements{trace, stateTVar, pool, securityParam} action = do
+handleChainIndexEffects
+        RunRequirements { trace, stateTVar, pool, securityParam = SecurityParam d }
+        action = do
     state <- liftIO $ readTVarIO stateTVar
     (result, newState) <-
         runState state
         $ runReader pool
-        $ runReader (Depth securityParam)
+        $ runReader (Depth $ fromIntegral d)
+        $ runReader chainIndexDb
         $ runError @ChainIndexError
         $ flip handleError (throwError . BeamEffectError)
         $ interpret (handleBeam (convertLog (PrettyObject . BeamLogItem) trace))
-        $ interpret handleControl
+
+        -- All indexers
+        $ handleChainIndexControl
+
         $ interpret handleQuery
-        -- Insert the 5 effects needed by the handlers of the 3 chain index effects between those 3 effects and 'effs'.
-        $ raiseMUnderN @[_,_,_,_,_] @[_,_,_] action
+
+        -- Insert the 6 effects needed by the handlers of the 3 chain index effects between those 3 effects and 'effs'.
+        $ raiseMUnderN @[_,_,_,_,_,_] @[_,_,_] action
     liftIO $ atomically $ writeTVar stateTVar newState
     pure result
+
+-- | Handle the chain index control effects from the set of all effects using the
+-- sqlite store.
+handleChainIndexControl :: forall effs.
+    ( Member BeamEffect effs
+    , Member (State ChainIndexState) effs
+    , Member (Reader Depth) effs
+    , Member (LogMsg ChainIndexLog) effs
+    , Member (Error ChainIndexError) effs
+    )
+    => Eff (ChainIndexControlEffect ': effs) ~> Eff effs
+handleChainIndexControl =
+    interpret
+        (\eff ->
+            handleUtxoIndexer chainIndexDb eff
+         >> handleDatumFromHashIndexer chainIndexDb eff
+         >> handleRedeemerFromHashIndexer chainIndexDb eff
+         >> handleScriptFromHashIndexer chainIndexDb eff
+         >> handleUtxoFromAddressIndexer chainIndexDb eff
+         >> handleAssetClassFromAddressIndexer chainIndexDb eff
```

Comment:
Hmmm, does it mean that all handlers should be run in one transaction?

If the app will be stopped in the middle of handlers, half of data will be written but half will be lost and we will get a corrupted index.

Reply:
Yes, it helped to combine all sql effects to execute in one transaction.